src/prefect/engine/flow_runner.py

```diff
@@ -717,6 +717,16 @@ def determine_final_state(
             self.logger.info("Flow run SUCCESS: no reference tasks failed")
             state = Success(message="No reference tasks failed.", result=return_states)
 
+        if self.flow.terminal_state_handler:
+            state = (
+                self.flow.terminal_state_handler(
+                    flow=self.flow,
+                    state=state,
+                    task_states=return_states,
+                )
+                or state
```

Comment:
Note this will do weird things if they return falsey values. I'm not sure it matters but an explicit `is not None` may expose mistakes as errors more clearly here.